nn/transformer.py

```diff
@@ -0,0 +1,320 @@
+               decoder_pos_enc: Callable = nn.positional_encoding, search: bool = False,
+               beam_size: Optional[int] = None) -> None:
+    """
+    :param output_dim: output dim, PyTorch name: d_model
+    :param num_heads: number heads, PyTorch name: nhead
+    :param num_encoder_layers: Number of encoder layers
+    :param num_decoder_layers: Number of decoder layers
+    :param dim_ff: dimension of feedforward layer, PyTorch name: dim_feedforward
+    :param dropout: Dropout value, PyTorch name: dropout
+    :param att_dropout: dropout value for attention
+    :param activation: activation function
+    :param custom_encoder: Custom Encoder to replace the standard encoder
+    :param custom_decoder: Custom Decoder to replace the standard decoder
+    :param custom_encoder_layer: Custom Encoder layer to replace the standard layer if custom_encoder and
+      custom_encoder_layer are given custom_encoder will be preferred
+    :param custom_decoder_layer: Custom Decoder layer to replace the standard layer if custom_decoder and
+      custom_decoder_layer are given custom_decoder will be preferred
+    :param norm_eps: Epsilon value for layer normalization
+    :param norm: function for layer normalization
+    :param encoder_pos_enc: Positional Encoding for the Encoder
+    :param decoder_pos_enc: Positional Encoding for the Decoder
+    :param search: Whether search is enabled or not
+    :param beam_size: Beam Size for search
+    """
+    super().__init__()
+
+    if custom_encoder is not None:
+      self.encoder = custom_encoder
+    else:
+      if custom_encoder_layer is not None:
+        encoder_layer = custom_encoder_layer
+      else:
+        encoder_layer = TransformerEncoderLayer(
+          dim_model=output_dim, num_heads=num_heads, dim_ff=dim_ff, dropout=dropout, activation=activation,
+          norm_eps=norm_eps, norm=norm, att_dropout=att_dropout)
+      self.encoder = TransformerEncoder(
+        encoder_layer=encoder_layer, num_layers=num_encoder_layers, norm=norm, norm_eps=norm_eps,
+        pos_enc=encoder_pos_enc)
+
+    if custom_decoder is not None:
+      self.decoder = custom_decoder
+    else:
+      if custom_decoder_layer is not None:
+        decoder_layer = custom_decoder_layer
+      else:
+        decoder_layer = TransformerDecoderLayerStep(
+          dim_model=output_dim, num_heads=num_heads, dim_ff=dim_ff, dropout=dropout, activation=activation,
+          norm_eps=norm_eps, norm=norm, att_dropout=att_dropout)
+      self.decoder = TransformerDecoderStep(
+        decoder_layer=decoder_layer, num_layers=num_decoder_layers, norm=norm, norm_eps=norm_eps,
+        pos_enc=decoder_pos_enc)
+
+    self.norm_eps = norm_eps
+    self.output_dim = output_dim
+    self.num_heads = num_heads
+    self.norm = norm
+    self.search = search
+    if search:
+      assert beam_size is not None
+      self.beam_size = beam_size
```

Comment:
You never should only conditionally assign an attrib. An attrib should always be there. Set `self.beam_size = None` otherwise or so.

But anyway, as explained before, beam size should not be here.